test/upgrade/deployment_failure.go

```diff
@@ -0,0 +1,134 @@
+/*
+Copyright 2024 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package upgrade
+
+import (
+	"context"
+
+	admissionv1 "k8s.io/api/admissionregistration/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"knative.dev/pkg/ptr"
+	"knative.dev/pkg/test/helpers"
+	pkgupgrade "knative.dev/pkg/test/upgrade"
+	"knative.dev/serving/pkg/apis/autoscaling"
+	v1 "knative.dev/serving/pkg/apis/serving/v1"
+	"knative.dev/serving/test"
+	"knative.dev/serving/test/e2e"
+	v1test "knative.dev/serving/test/v1"
+)
+
+func DeploymentFailurePostUpgrade() pkgupgrade.Operation {
+	return pkgupgrade.NewOperation("DeploymentFailurePostUpgrade", func(c pkgupgrade.Context) {
+		clients := e2e.Setup(c.T)
+
+		names := test.ResourceNames{
+			Service: "deployment-upgrade-failure",
+			Image:   test.HelloWorld,
+		}
+
+		service, err := clients.ServingClient.Services.Get(context.Background(), names.Service, metav1.GetOptions{})
+		if err != nil {
+			c.T.Fatal("Failed to get Service: ", err)
+		}
+
+		// Deployment failures should surface to the Service Ready Condition
+		if err := v1test.WaitForServiceState(clients.ServingClient, names.Service, v1test.IsServiceFailed, "ServiceIsNotReady"); err != nil {
+			c.T.Fatal("Service did not transition to Ready=False", err)
+		}
+
+		// Traffic should still work since the deployment has an active replicaset
+		url := service.Status.URL.URL()
+		assertServiceResourcesUpdated(c.T, clients, names, url, test.HelloWorldText)
+	})
+}
+
+func DeploymentFailurePreUpgrade() pkgupgrade.Operation {
+	return pkgupgrade.NewOperation("DeploymentFailurePreUpgrade", func(c pkgupgrade.Context) {
+		c.T.Log("Creating Service")
+		ctx := context.Background()
+
+		clients := e2e.Setup(c.T)
+		names := &test.ResourceNames{
+			Service: "deployment-upgrade-failure",
+			Image:   test.HelloWorld,
+		}
+
+		resources, err := v1test.CreateServiceReady(c.T, clients, names, func(s *v1.Service) {
+			s.Spec.Template.Annotations = map[string]string{
+				autoscaling.MinScaleAnnotation.Key(): "1",
+				autoscaling.MaxScaleAnnotation.Key(): "1",
```

Comment:
What will happen if maxScale = 10 and we deploy the failing webhook before all replicas are up? Would the new revision be reachable since some replicas are up?

Reply:
The replica set has techinically progressed so there would be no failure surfaced on the deployment because it's a scaling issue for the replicaset.